formats/json/commonMain/src/kotlinx/serialization/json/internal/StreamingJsonEncoder.kt

```diff
@@ -160,10 +160,18 @@ internal class StreamingJsonEncoder(
 
     override fun encodeInline(inlineDescriptor: SerialDescriptor): Encoder =
         if (inlineDescriptor.isUnsignedNumber) StreamingJsonEncoder(
-            ComposerForUnsignedNumbers(composer.sb), json, mode, null
+            composerForUnsignedNumbers(), json, mode, null
         )
         else super.encodeInline(inlineDescriptor)
 
+    private fun composerForUnsignedNumbers(): Composer {
+        // If we're inside encodeInline().encodeSerializableValue, we should preserve the forceQuoting state
+        // inside the composer, but not in the encoder (otherwise we'll get into `if (forceQuoting) encodeString(value.toString())` part
+        // and unsigned numbers would be encoded incorrectly)
+        return if (composer is ComposerForUnsignedNumbers) composer
```

Comment:
I can lift `forceQuoting` to the base `Composer` class to avoid is-instance check. Although I'm not sure this micro-optimization would make any difference.